waiter/src/waiter/utils.clj

```diff
@@ -183,22 +186,91 @@
+  (let [accept (or (get headers "accept") "text/plain")]
+    (cond
+      (str/includes? accept "application/json") "application/json"
+      (str/includes? accept "text/html") "text/html"
+      :else "text/plain")))
+
+(defn error-context->text
+  "Formats an error context for a text response."
+  [{:keys [cid details host message query-string request-method status timestamp uri]}]
+  (let [padded-format (fn [f v] (format "%16s: %s\n" f v))]
+    (str "\n" 
+         "  Waiter Error " status "\n" 
+         "  ================\n\n"
+         "    " message "\n\n"
+         "  Request Info\n" 
+         "  ============\n\n"
+         (when host (padded-format "Host" host))
+         (padded-format "Path" uri)
+         (when query-string (padded-format "Query String" query-string))
+         (padded-format "Method" request-method)
+         (padded-format "CID" cid)
+         (padded-format "Time" timestamp)
+         (when (seq details) 
+           (str "\n"
+                "  Additional Info\n" 
+                "  ===============\n\n   "
+                (-> (with-out-str 
+                      (pprint/pprint details))
+                    (str/replace #"\n" "\n    ")) 
+                "\n\n")))))
+
 (defn exception->response
-  "Converts an exception into a 400 plain text response with the stack trace."
-  [log-message ^Exception e & {:keys [headers status] :or {headers {}}}]
+  "Converts an exception into a 400 response with an error message."
+  [request log-message ^Exception e & {:keys [headers status] :or {headers {}}}]
   (let [processed-headers (into {} (for [[k v] headers] [(name k) (str v)]))
         ex-data (ex-data e)]
     (when-not (:suppress-logging ex-data)
       (log/error e log-message processed-headers))
-    {:body (:message ex-data (str/join (System/lineSeparator) (exception->strs e)))
-     :status (exception->status e :status status)
-     :headers (merge {"Content-Type" "text/plain"} processed-headers)}))
+    (let [message (or (:friendly-error-message ex-data)
+                      (:message ex-data)
+                      (.getMessage e))
+          content-type (request->content-type request)
+          timestamp (date-to-str (t/now))
+          cid (-> request :headers (get "x-cid"))
+          host (-> request :headers (get "host"))
+          request-method (-> request :request-method (or "") name str/upper-case)
+          query-string (-> request :query-string)
+          uri (-> request :uri)
+          status (or status (exception->status e))
+          error-context {:cid cid 
+                         :details ex-data
+                         :host host
+                         :message message 
+                         :query-string query-string
+                         :request-method request-method
+                         :status status
+                         :timestamp timestamp
+                         :uri uri}]
+      {:status status
+       :body (case content-type
```

Comment:
I think this case needs to be in its own function and we can make `request->content-type` redundant. It also saves us having to repeat the case statement for the three types.

Reply:
There are two separate concerns here.

1. Determining what content-type is appropriate.
2. Formatting output based upon content-type.

They shouldn't be the same function.